libopenage/renderer/font.cpp

```diff
@@ -0,0 +1,227 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#include "font.h"
+
+#include <ft2build.h>
+#include FT_FREETYPE_H
+#include <harfbuzz/hb.h>
+#include <harfbuzz/hb-ft.h>
+
+#include "../error/error.h"
+#include "../util/compiler.h"
+
+#include "font_manager.h"
+
+namespace openage {
+namespace renderer {
+
+font_description::font_description(const std::string &font_file,
+                                   unsigned int size,
+                                   font_direction direction,
+                                   const std::string &language,
+                                   const std::string &script)
+	:
+	font_file{font_file},
+	size{size},
+	direction{direction},
+	language{language},
+	script{script} {
+	// Empty
+}
+
+font_description::font_description(const char *family, const char *style, unsigned int size)
+	:
+	font_description{FontManager::get_font_filename(family, style), size} {
+	// Empty
+}
+
+font_description &font_description::operator=(const font_description &other) {
+	if (this != &other) {
+		this->font_file = other.font_file;
+		this->size = other.size;
+		this->direction = other.direction;
+		this->language = other.language;
+		this->script = other.script;
+	}
+
+	return *this;
+}
+
+font_description::font_description(const font_description &other)
+	:
+	font_file{other.font_file},
+	size{other.size},
+	direction{other.direction},
+	language{other.language},
+	script{other.script} {
+	// Empty
+}
+
+bool font_description::operator==(const font_description &other) const {
+	return this->font_file == other.font_file &&
+		this->size == other.size &&
+		this->direction == other.direction &&
+		this->language == other.language &&
+		this->script == other.script;
+}
+
+bool font_description::operator!=(const font_description &other) const {
+	return !(*this == other);
+}
+
+size_t font_description::hash() const {
+	std::size_t h1 = std::hash<std::string>()(this->font_file);
+	std::size_t h2 = std::hash<unsigned int>()(this->size);
+	// Taken from http://www.boost.org/doc/libs/1_55_0/doc/html/hash/reference.html#boost.hash_combine
+	return h1 ^ (h2 + 0x9e3779b9 + ((h1 << 6) + (h1 >> 2)));
```

Comment:
Hm you seem to be right. I'd suggest adding a function in `util` to allow combining hashes, so you start with some seed and can update it again and again with any another hash (a `size_t` number). Then i'd add `std::hash<font_description>` that uses that hash combination function. Better than duplicating the combination algorithm all over the codebase.
The existing combinations in some other parts of the code should be updated as well then.